packages/postgres-database/src/simcore_postgres_database/utils_projects_nodes.py

```diff
@@ -0,0 +1,215 @@
+                # this happens if the node already exists
+                raise ProjectsNodesDuplicateNode(
+                    f"Project node {node.node_id} already exists"
+                ) from exc
+
+    async def add(
+        self, connection: SAConnection, *, node_id: uuid.UUID
+    ) -> ProjectsNode:
+        """adds a node with node_id to the current project
+
+        NOTE: Do not use this in an asyncio.gather call as this will fail!
+
+        Raises:
+            ProjectsNodesOperationNotAllowed: _description_
+        """
+        try:
+            result = await connection.execute(
+                projects_to_projects_nodes.insert().values(
+                    project_uuid=f"{self.project_uuid}",
+                    node_id=f"{node_id}",
+                )
+            )
+            assert result.rowcount == 1  # nosec
+
+            return await self.get(connection, node_id=node_id)
+
+        except ForeignKeyViolation as exc:
+            raise ProjectsNodesOperationNotAllowed(
+                f"Node {node_id=} cannot be added to project {self.project_uuid}"
+            ) from exc
+
+    async def list(self, connection: SAConnection) -> list[ProjectsNode]:
+        """list the nodes in the current project
+
+        NOTE: Do not use this in an asyncio.gather call as this will fail!
+        """
+        list_stmt = (
+            sqlalchemy.select(projects_nodes)
+            .select_from(self._join_projects_to_projects_nodes())
+            .where(projects_to_projects_nodes.c.project_uuid == f"{self.project_uuid}")
+        )
+        nodes = [
+            ProjectsNode(**dict(row.items()))
+            async for row in connection.execute(list_stmt)
+        ]
+        return nodes
+
+    async def get(
+        self, connection: SAConnection, *, node_id: uuid.UUID
+    ) -> ProjectsNode:
+        """get a node in the current project
+
+        NOTE: Do not use this in an asyncio.gather call as this will fail!
+
+        Raises:
+            ProjectsNodesNodeNotFound: _description_
+        """
+
+        get_stmt = (
+            sqlalchemy.select(projects_nodes)
+            .select_from(self._join_projects_to_projects_nodes())
+            .where(
+                (projects_to_projects_nodes.c.project_uuid == f"{self.project_uuid}")
+                & (projects_to_projects_nodes.c.node_id == f"{node_id}")
+            )
+        )
+
+        result = await connection.execute(get_stmt)
+        assert result  # nosec
+        row = await result.first()
+        if row is None:
+            raise ProjectsNodesNodeNotFound(f"Node with {node_id} not found")
+        assert row  # nosec
+        return ProjectsNode(**dict(row.items()))
+
+    @staticmethod
+    async def update(
+        connection: SAConnection, *, node_id: uuid.UUID, **values
+    ) -> ProjectsNode:
+        """update a node in the current project
+
+        NOTE: Do not use this in an asyncio.gather call as this will fail!
```

Comment:
Some consideration as above. This can be triggered by the fronted making multiple requests in parallel resulting in the same situation as asyncio.gather

Reply:
discussed and solved.